File: sort-tool/rust_sort_tool/src/algorithms/merge_sort.rs
pub fn sort(arr: &mut [String]) {
    if arr.len() > 1 {
        let mid = arr.len() / 2;
        sort(&mut arr[..mid]);
        sort(&mut arr[mid..]);
        merge(arr, mid);
    }
}

fn merge(arr: &mut [String], mid: usize) {
    let left = arr[..mid].to_vec();
    let right = arr[mid..].to_vec();

    let mut i = 0;
    let mut j = 0;
    let mut k = 0;

    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            arr[k] = left[i].clone();
            i += 1;
        } else {
            arr[k] = right[j].clone();
            j += 1;
        }
        k += 1;
    }

    while i < left.len() {
        arr[k] = left[i].clone();
        i += 1;
        k += 1;
    }

    while j < right.len() {
        arr[k] = right[j].clone();
        j += 1;
        k += 1;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test] 
    fn test_merge_sort() {
        let mut input = vec![
            "banana".to_string(),
            "apple".to_string(),
            "cherry".to_string(),
            "date".to_string(),
            "elderberry".to_string(),
        ];

        let expected = vec![
            "apple".to_string(),
            "banana".to_string(),
            "cherry".to_string(),
            "date".to_string(),
            "elderberry".to_string(),
        ];

        sort(&mut input);

        assert_eq!(input, expected);
    }
}

File: sort-tool/go_sort_tool/sorter/algorithms/merge_sort.go
package algorithms

func MergeSort(arr []string) {
    if len(arr) <= 1 {
        return 
    }

    mid := len(arr) / 2
    left := make([]string, mid)
    right := make([]string, len(arr)-mid)

    copy(left, arr[:mid])
    copy(right, arr[mid:])

    MergeSort(left)
    MergeSort(right)

    merge(arr, left, right)
}

func merge(arr, left, right []string) {
    i, j, k := 0, 0, 0

    for i < len(left) && j < len(right) {
        if left[i] <= right[j] {
            arr[k] = left[i]
            i++
        } else {
            arr[k] = right[j]
            j++
        }
        k++
    }

    for i < len(left) {
        arr[k] = left[i]
        i++
        k++
    }

    for j < len(right) {
        arr[k] = right[j]
        j++
        k++
    }
}
